go-controller/pkg/node/gateway_shared_intf_linux.go

```diff
@@ -71,10 +72,20 @@ func setupLocalNodeAccessBridge(nodeName string, subnets []*net.IPNet) error {
 			gatewaySubnetMask = net.CIDRMask(util.V4NodeLocalNatSubnetPrefix, 32)
 		}
 		gatewayNextHopCIDR := &net.IPNet{IP: gatewayNextHop, Mask: gatewaySubnetMask}
-
 		if err = util.LinkAddrAdd(link, gatewayNextHopCIDR); err != nil {
 			return err
 		}
+		gatewayIfAddrs = append(gatewayIfAddrs, gatewayNextHopCIDR)
+	}
+
+	if config.Gateway.Mode == config.GatewayModeLocal {
+		// need to add masquerading for ovn-k8s-gw0 port for hostA -> service -> hostB via DGP
```

Comment:
Note: I'm not really satisfied with how the DGP solution works, since we end up doing a circular traffic loop to get host-originated traffic back into the host like:

host -> service -> mp0 into ovn-> forward to gw0->back into host

Which involves a couple SNAT's to trick the host into thinking traffic is from external networks, even though it came form his own host and came back in another interface.

This is the current way we handle service traffic for host backed services today, so I'm leveraging it here. I have some ideas we can talk about in the next community meeting on how to improve this and possibly get rid of the DGP and gw0 port altogether in OVN, but I'd like to refrain from discussing it in this PR and handle it in the community meeting.

Reply:
I honestly think we should kill mp0 entirely and make all the traffic go through the gateway, becuase then we have better control over the traffic due to the routers. It's a problem that mp0 is direct to the node switch, and we must rely on ARP/L2 + SNAT to get traffic back out mp0 when needed. There's no good reason to rely on L2 for any of this.

Reply:
@dcbw I don't see the problem. We have control over egress traffic because pods always send (non-svc traffic) to ovn_cluster_router first, that means egress traffic is always routed. For ingress traffic, its always svc traffic, which gets load balanced on the switch anyway. I'd like to kill gw0 instead, but this is a discussion for community meeting I think.